### PR TITLE
bugfix:fix wrong exception information when rollback fails due to dirty data

### PR DESCRIPTION
**Pull request by a364176773**
### Ⅰ. Describe what this PR did

fix wrong exception information when rollback fails due to dirty data

### Ⅱ. Does this pull request fix one issue?
#2325

### CODE REVIEW

**Comment by jovany-wang** [comment]
# [Codecov](https://codecov.io/gh/seata/seata/pull/2333?src=pr&el=h1) Report
> Merging [#2333](https://codecov.io/gh/seata/seata/pull/2333?src=pr&el=desc) into [develop](https://codecov.io/gh/seata/seata/commit/83393442e8a8c7b048d10b104a5018c0fea1e045&el=desc) will **decrease** coverage by `0.15%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/seata/seata/pull/2333?src=pr&el=tree)

```diff
@@              Coverage Diff              @@
##             develop    #2333      +/-   ##
=============================================
- Coverage      51.61%   51.46%   -0.16%     
  Complexity      2691     2691              
=============================================
  Files            517      517              
  Lines          16728    16711      -17     
  Branches        2038     2000      -38     
=============================================
- Hits            8634     8600      -34     
+ Misses          7284     7278       -6     
- Partials         810      833      +23     
```


| [Impacted Files](https://codecov.io/gh/seata/seata/pull/2333?src=pr&el=tree) | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| [...obuf/convertor/BranchRegisterRequestConvertor.java](https://codecov.io/gh/seata/seata/pull/2333/diff?src=pr&el=tree#diff-c2VyaWFsaXplci9zZWF0YS1zZXJpYWxpemVyLXByb3RvYnVmL3NyYy9tYWluL2phdmEvaW8vc2VhdGEvc2VyaWFsaXplci9wcm90b2J1Zi9jb252ZXJ0b3IvQnJhbmNoUmVnaXN0ZXJSZXF1ZXN0Q29udmVydG9yLmphdmE=) | `90.47% <0.00%> (-9.53%)` | `3.00% <0.00%> (ø%)` | |
| [...otobuf/convertor/BranchReportRequestConvertor.java](https://codecov.io/gh/seata/seata/pull/2333/diff?src=pr&el=tree#diff-c2VyaWFsaXplci9zZWF0YS1zZXJpYWxpemVyLXByb3RvYnVmL3NyYy9tYWluL2phdmEvaW8vc2VhdGEvc2VyaWFsaXplci9wcm90b2J1Zi9jb252ZXJ0b3IvQnJhbmNoUmVwb3J0UmVxdWVzdENvbnZlcnRvci5qYXZh) | `90.90% <0.00%> (-9.10%)` | `3.00% <0.00%> (ø%)` | |
| [...obuf/convertor/BranchRollbackRequestConvertor.java](https://codecov.io/gh/seata/seata/pull/2333/diff?src=pr&el=tree#diff-c2VyaWFsaXplci9zZWF0YS1zZXJpYWxpemVyLXByb3RvYnVmL3NyYy9tYWluL2phdmEvaW8vc2VhdGEvc2VyaWFsaXplci9wcm90b2J1Zi9jb252ZXJ0b3IvQnJhbmNoUm9sbGJhY2tSZXF1ZXN0Q29udmVydG9yLmphdmE=) | `92.00% <0.00%> (-8.00%)` | `3.00% <0.00%> (ø%)` | |
| [...otobuf/convertor/GlobalBeginResponseConvertor.java](https://codecov.io/gh/seata/seata/pull/2333/diff?src=pr&el=tree#diff-c2VyaWFsaXplci9zZWF0YS1zZXJpYWxpemVyLXByb3RvYnVmL3NyYy9tYWluL2phdmEvaW8vc2VhdGEvc2VyaWFsaXplci9wcm90b2J1Zi9jb252ZXJ0b3IvR2xvYmFsQmVnaW5SZXNwb25zZUNvbnZlcnRvci5qYXZh) | `92.59% <0.00%> (-7.41%)` | `3.00% <0.00%> (ø%)` | |
| [...otobuf/convertor/GlobalCommitRequestConvertor.java](https://codecov.io/gh/seata/seata/pull/2333/diff?src=pr&el=tree#diff-c2VyaWFsaXplci9zZWF0YS1zZXJpYWxpemVyLXByb3RvYnVmL3NyYy9tYWluL2phdmEvaW8vc2VhdGEvc2VyaWFsaXplci9wcm90b2J1Zi9jb252ZXJ0b3IvR2xvYmFsQ29tbWl0UmVxdWVzdENvbnZlcnRvci5qYXZh) | `93.75% <0.00%> (-6.25%)` | `3.00% <0.00%> (ø%)` | |
| [...otobuf/convertor/GlobalStatusRequestConvertor.java](https://codecov.io/gh/seata/seata/pull/2333/diff?src=pr&el=tree#diff-c2VyaWFsaXplci9zZWF0YS1zZXJpYWxpemVyLXByb3RvYnVmL3NyYy9tYWluL2phdmEvaW8vc2VhdGEvc2VyaWFsaXplci9wcm90b2J1Zi9jb252ZXJ0b3IvR2xvYmFsU3RhdHVzUmVxdWVzdENvbnZlcnRvci5qYXZh) | `93.75% <0.00%> (-6.25%)` | `3.00% <0.00%> (ø%)` | |
| [...obuf/convertor/GlobalRollbackRequestConvertor.java](https://codecov.io/gh/seata/seata/pull/2333/diff?src=pr&el=tree#diff-c2VyaWFsaXplci9zZWF0YS1zZXJpYWxpemVyLXByb3RvYnVmL3NyYy9tYWluL2phdmEvaW8vc2VhdGEvc2VyaWFsaXplci9wcm90b2J1Zi9jb252ZXJ0b3IvR2xvYmFsUm9sbGJhY2tSZXF1ZXN0Q29udmVydG9yLmphdmE=) | `93.75% <0.00%> (-6.25%)` | `3.00% <0.00%> (ø%)` | |
| [...protobuf/convertor/RegisterTMRequestConvertor.java](https://codecov.io/gh/seata/seata/pull/2333/diff?src=pr&el=tree#diff-c2VyaWFsaXplci9zZWF0YS1zZXJpYWxpemVyLXByb3RvYnVmL3NyYy9tYWluL2phdmEvaW8vc2VhdGEvc2VyaWFsaXplci9wcm90b2J1Zi9jb252ZXJ0b3IvUmVnaXN0ZXJUTVJlcXVlc3RDb252ZXJ0b3IuamF2YQ==) | `94.73% <0.00%> (-5.27%)` | `3.00% <0.00%> (ø%)` | |
| [...protobuf/convertor/RegisterRMRequestConvertor.java](https://codecov.io/gh/seata/seata/pull/2333/diff?src=pr&el=tree#diff-c2VyaWFsaXplci9zZWF0YS1zZXJpYWxpemVyLXByb3RvYnVmL3NyYy9tYWluL2phdmEvaW8vc2VhdGEvc2VyaWFsaXplci9wcm90b2J1Zi9jb252ZXJ0b3IvUmVnaXN0ZXJSTVJlcXVlc3RDb252ZXJ0b3IuamF2YQ==) | `90.47% <0.00%> (-4.77%)` | `3.00% <0.00%> (ø%)` | |
| [...tobuf/convertor/BranchReportResponseConvertor.java](https://codecov.io/gh/seata/seata/pull/2333/diff?src=pr&el=tree#diff-c2VyaWFsaXplci9zZWF0YS1zZXJpYWxpemVyLXByb3RvYnVmL3NyYy9tYWluL2phdmEvaW8vc2VhdGEvc2VyaWFsaXplci9wcm90b2J1Zi9jb252ZXJ0b3IvQnJhbmNoUmVwb3J0UmVzcG9uc2VDb252ZXJ0b3IuamF2YQ==) | `95.45% <0.00%> (-4.55%)` | `3.00% <0.00%> (ø%)` | |
| ... and [19 more](https://codecov.io/gh/seata/seata/pull/2333/diff?src=pr&el=tree-more) | |

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java [hunk 1]
```diff
@@ -143,7 +144,9 @@ private void rollbackTransaction(GlobalTransaction tx, Throwable ex) throws Tran
         tx.rollback();
         triggerAfterRollback();
         // 3.1 Successfully rolled back
-        throw new TransactionalExecutor.ExecutionException(tx, TransactionalExecutor.Code.RollbackDone, ex);
+        throw new TransactionalExecutor.ExecutionException(tx,
```
Comment (slievrly) [inline tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: I feel that the code you changed is not correct. I think it is OK to throw RollbackDone directly here.

**Comment by slievrly** [comment]
@a364176773 check the code, no problem. However, there are a few minor problems. Tx.getStatus () is a remote call, but in fact status is already available when rollback. The second is that the exception thrown by tx.rollback () must be the rpc exception of the framework and the business rollback fails Both are defined with RollbackFailure, without distinction.

**Comment by slievrly** [comment]
@a364176773 Please solve the problem of getStatus to get the remote status, in fact status has been returned in rollback.

**Comment by a364176773** [comment]
> @a364176773 Please solve the problem of getStatus to get the remote status, in fact status has been returned in rollback.

@slievrly i will submit in additional

**Comment by a364176773** [comment]
@zjinlei @xingfudeshi PTAL

**Comment by a364176773** [comment]
@xingfudeshi @zjinlei PTAL

File: tm/src/main/java/io/seata/tm/api/DefaultFailureHandlerImpl.java [hunk 2]
```diff
@@ -67,6 +68,14 @@ public void onRollbackFailure(GlobalTransaction tx, Throwable cause) {
         timer.newTimeout(new CheckTimerTask(tx, GlobalStatus.Rollbacked), SCHEDULE_INTERVAL_SECONDS, TimeUnit.SECONDS);
     }
 
+    @Override
+    public void onRollbackRetrying(GlobalTransaction tx, Throwable cause) {
+        StackTraceLogger.info(LOGGER, cause, "Retrying to rollback transaction[{}]", new String[] {tx.getXid()}, null,
+            null);
```
Comment (zjinlei) [inline tm/src/main/java/io/seata/tm/api/DefaultFailureHandlerImpl.java]: StackTraceLogger.info(LOGGER, cause, "Retrying to rollback transaction[{}]", new String[] {tx.getXid()}, "Retrying to rollback transaction[{}]", new String[] {tx.getXid()});

Comment (a364176773) [reply tm/src/main/java/io/seata/tm/api/DefaultFailureHandlerImpl.java]: > StackTraceLogger.info(LOGGER, cause, "Retrying to rollback transaction[{}]", new String[] {tx.getXid()}, "Retrying to rollback transaction[{}]", new String[] {tx.getXid()});

PTAL @zjinlei

File: tm/src/main/java/io/seata/tm/api/DefaultFailureHandlerImpl.java [hunk 3]
```diff
@@ -67,6 +67,12 @@ public void onRollbackFailure(GlobalTransaction tx, Throwable cause) {
         timer.newTimeout(new CheckTimerTask(tx, GlobalStatus.Rollbacked), SCHEDULE_INTERVAL_SECONDS, TimeUnit.SECONDS);
     }
 
+    @Override
+    public void onRollbackRetrying(GlobalTransaction tx, Throwable cause) {
+        LOGGER.warn("Failed to rollback transaction[" + tx.getXid() + "]", cause);
```
[the remark this answers is not in the document]
Comment (a364176773) [reply tm/src/main/java/io/seata/tm/api/DefaultFailureHandlerImpl.java]: @xingfudeshi PTAL

**Pull request merged**